### PR TITLE
A potential optimization to NetUtil#getIPv6ByName

### PR DESCRIPTION
**Pull request by lijinliangyihao**
I ran the related tests and they all passed.

Result: potential efficiency increment

see #12107

### CODE REVIEW

**Comment by chrisvest** [comment]
@lijinliangyihao Please sign the ICLA: https://netty.io/wiki/developer-guide.html

**Comment by hyperxpro** [comment]
The benchmark code is very unreadable. Can you simplify it better?

**Comment by lijinliangyihao** [comment]
> The benchmark code is very unreadable. Can you simplify it better?

Sorry, I need some explanation. I modified a small part of the method, NetUtil#getIPv6ByName, and I have to prove my modification increases performance.

There're two methods in a new class named `GetByNameHolder`:
`getIPv6ByNameNew` - which is the version I modified, and `getIPv6ByNameOld` - the original version.

Here's the benchmark on my machine, windows i7 6c 16g:
```
Benchmark                                                          (ip)   Mode  Cnt   Score   Error   Units
GetByIpV6NameBenchmark.getNameNew                                    ::  thrpt    6  73.737 ± 9.567  ops/us
GetByIpV6NameBenchmark.getNameNew                           ::1234:2345  thrpt    6  17.278 ± 1.200  ops/us
GetByIpV6NameBenchmark.getNameNew                  1234:2345::3456:7890  thrpt    6  10.421 ± 1.756  ops/us
GetByIpV6NameBenchmark.getNameNew             fe80::bdad:7a67:6fcd:fa89  thrpt    6   7.026 ± 1.779  ops/us
GetByIpV6NameBenchmark.getNameNew             fe80:bdad:7a67:6fcd::fa89  thrpt    6   8.287 ± 2.483  ops/us
GetByIpV6NameBenchmark.getNameNew  1234:2345:3456:4567:5678:6789:0:7890  thrpt    6   7.479 ± 1.209  ops/us

GetByIpV6NameBenchmark.getNameOld                                    ::  thrpt    6  20.109 ± 5.807  ops/us
GetByIpV6NameBenchmark.getNameOld                           ::1234:2345  thrpt    6   8.592 ± 6.492  ops/us
GetByIpV6NameBenchmark.getNameOld                  1234:2345::3456:7890  thrpt    6   5.300 ± 2.342  ops/us
GetByIpV6NameBenchmark.getNameOld             fe80::bdad:7a67:6fcd:fa89  thrpt    6   5.592 ± 2.373  ops/us
GetByIpV6NameBenchmark.getNameOld             fe80:bdad:7a67:6fcd::fa89  thrpt    6   6.135 ± 2.234  ops/us
GetByIpV6NameBenchmark.getNameOld  1234:2345:3456:4567:5678:6789:0:7890  thrpt    6   6.318 ± 1.445  ops/us
```

We can see, the shorter the ip is, the more performance increasement we gain.

So why copy the new code from NetUtil, rather than just use NetUtil? Because I tried many times, if I use NetUtil's code directly, for ip '1234:2345:3456:4567:5678:6789:0:7890', NetUtil's code is alway a little bit slower than GetByNameHolder's(using the same code). So in order to eliminate noise, I just copy the code from NetUtil to GetByNameHolder, and the benchmark just compare the `getNameNew` and `getNameOld`.

**Comment by lijinliangyihao** [comment]
If I use code in NetUtil(modified version), instead of copy the code to the class `GetByNameHolder`, here's the benchmark result:

```
Benchmark                                                          (ip)   Mode  Cnt   Score    Error   Units
GetByIpV6NameBenchmark.getNameNew                                    ::  thrpt    6  40.745 ± 11.307  ops/us
GetByIpV6NameBenchmark.getNameNew                           ::1234:2345  thrpt    6  14.041 ±  4.927  ops/us
GetByIpV6NameBenchmark.getNameNew                  1234:2345::3456:7890  thrpt    6   8.942 ±  2.955  ops/us
GetByIpV6NameBenchmark.getNameNew             fe80::bdad:7a67:6fcd:fa89  thrpt    6   6.351 ±  1.598  ops/us
GetByIpV6NameBenchmark.getNameNew             fe80:bdad:7a67:6fcd::fa89  thrpt    6   6.421 ±  2.446  ops/us
GetByIpV6NameBenchmark.getNameNew  1234:2345:3456:4567:5678:6789:0:7890  thrpt    6   5.552 ±  1.566  ops/us

GetByIpV6NameBenchmark.getNameOld                                    ::  thrpt    6  16.704 ±  5.078  ops/us
GetByIpV6NameBenchmark.getNameOld                           ::1234:2345  thrpt    6   7.991 ±  3.059  ops/us
GetByIpV6NameBenchmark.getNameOld                  1234:2345::3456:7890  thrpt    6   6.323 ±  1.685  ops/us
GetByIpV6NameBenchmark.getNameOld             fe80::bdad:7a67:6fcd:fa89  thrpt    6   6.223 ±  0.137  ops/us
GetByIpV6NameBenchmark.getNameOld             fe80:bdad:7a67:6fcd::fa89  thrpt    6   6.819 ±  0.718  ops/us
GetByIpV6NameBenchmark.getNameOld  1234:2345:3456:4567:5678:6789:0:7890  thrpt    6   6.830 ±  0.110  ops/us
```

As we can see, when ip is short, new version is still faster than original version, but for the last input param, new version is slower. That does not make sense! Because my modification has no effect if ip is not compressed, so the result should be very close.

Here's another comparison. I tried 3 or 4 time, the result is similar.
```
Benchmark                                                          (ip)   Mode  Cnt  Score   Error   Units
GetByIpV6NameBenchmark.getNameNew  1234:2345:3456:4567:5678:6789:0:7890  thrpt    6  6.691 ± 0.699  ops/us
GetByIpV6NameBenchmark.getNameOld  1234:2345:3456:4567:5678:6789:0:7890  thrpt    6  7.875 ± 0.303  ops/us
```

For this reason, I copy the code from NetUtil to the new helper class.

File: microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,448 @@
+/*
+ * Copyright 2017 The Netty Project
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java:2]:
```suggestion
 * Copyright 2022 The Netty Project
```

File: microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        "fe80::bdad:7a67:6fcd:fa89",
+        "fe80:bdad:7a67:6fcd::fa89",
+        "1234:2345:3456:4567:5678:6789:0:7890"
+    })
+    private String ip;
+
+    @Benchmark
+    public void getNameOld() {
+        GetByNameHolder.getIPv6ByNameOld(ip, true);
+    }
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java:51]:
```suggestion
    public byte[] getNameOld() {
        return GetByNameHolder.getIPv6ByNameOld(ip, true);
    }
```

File: microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+
+    @Benchmark
+    public void getNameNew() {
+        GetByNameHolder.getIPv6ByNameNew(ip, true);
+    }
```
Comment (normanmaurer) [inline microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java:56]:
```suggestion
    public byte[] getNameNew() {
        return GetByNameHolder.getIPv6ByNameNew(ip, true);
    }
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
you need to return something from the benchmark or use a blackhole.

File: microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+                ipv6Separators < IPV6_MIN_SEPARATORS ||
+                !isCompressed && (ipv6Separators + 1 != IPV6_MAX_SEPARATORS  ||
+                    ip.charAt(0) == ':' || ip.charAt(tmp) == ':') ||
+                isCompressed && (ipv6Separators > IPV6_MAX_SEPARATORS ||
+                    (ipv6Separators == IPV6_MAX_SEPARATORS &&
+                        (compressBegin <= 2 && ip.charAt(0) != ':' ||
+                            compressBegin >= 14 && ip.charAt(tmp) != ':'))) ||
+                currentIndex + 1 >= bytes.length ||
+                begin < 0 && ip.charAt(tmp - 1) != ':' ||
+                compressBegin > 2 && ip.charAt(0) == ':') {
```
[the remark this answers is not in the document]
Comment (lijinliangyihao) [reply microbench/src/main/java/io/netty/microbenchmark/common/GetByIpV6NameBenchmark.java]: I didn't modify this part of the original code. 
I committed a new version of the benchmark code for better understanding, please have a look.

**Comment by lijinliangyihao** [comment]
> Please add a jmh benchmark to validate the improvement

Benchmark shows about 17% ~ 265% throughput increasement.

**Comment by franz1981** [comment]
I still think there's lot of room of improvement up there, but still..this is already a small change that deliver a nice improvement.
I didn't checked if functionally things still work the same, so this is just a performance-focused-review TLDR well done @lijinliangyihao 

Just a quick question: how do you spot it? it doesn't seems a method that's used on an hot path (and I could be wrong!)

**Comment by lijinliangyihao** [comment]
> I still think there's lot of room of improvement up there, but still..this is already a small change that deliver a nice improvement. I didn't checked if functionally things still work the same, so this is just a performance-focused-review TLDR well done @lijinliangyihao
> 
> Just a quick question: how do you spot it? it doesn't seems a method that's used on an hot path (and I could be wrong!)

I'm reading the code of netty recently, and see if I can make some contribution to it.

**Review by chrisvest — APPROVED** [review]
👍

**Comment by normanmaurer** [comment]
@lijinliangyihao thanks a lot!

**Pull request merged**